Add option to specify table env. in LaTeX output

pajowu opened the pull request:
This PR adds a command line switch (`--latex-table-environment`) for switching between possible table environments. Currently pandoc is hardcoded to use `longtable`, which allows for multi-page-tables. This adds support for the tabularx environment (allowing columns that span over the while remaining table-width) and gives the possibility to easily add new environments in the future.

**NOTE**: The package used in the template could also be `ltablex`, which is a combination of longtable and tabularx (allows multi-page tables with auto-spanning columns).

pajowu commented:
Is there something blocking this PR?

jgm commented on MANUAL.txt:
```diff
@@ -1054,8 +1054,16 @@ Options affecting specific writers
+`--latex-table-environment=longtable`|`tabularx`
+
+:   Use the specified table environment when producing LaTeX output.
+    The default is [`longtable`] which allows table to spread over multiple tables.
```
Do you mean "over multiple pages"?

jgm commented on MANUAL.txt:
```diff
@@ -1054,8 +1054,16 @@ Options affecting specific writers
+
+:   Use the specified table environment when producing LaTeX output.
+    The default is [`longtable`] which allows table to spread over multiple tables.
+    [`tabularx`] allows for columns which autospread over the entire linewidth.
```
I wasn't sure exactly what was meant by "autospread over the entire linewidth."

jgm commented:
Sorry, I just haven't had time to review it.  Can you say more about the motivation for the change?

themightyoarfish commented:
One motivation from my perspective is that `longtable` is broken in the presence of floats on the same page (see e.g.[here](https://tex.stackexchange.com/questions/326882/figure-image-is-overflowing-bottom-of-page) or [here](https://tex.stackexchange.com/questions/276699/the-longtable-environment-pushes-content-below-it-into-the-bottom-margin-of-a-pa#277542))

In my case, this happily pushes sections into the bottom margin and the space between pages, so they are lost. 

This is a known bug in `longtable`, but unfortunately it also breaks pandoc use in cases, so an option for this would be great.

themightyoarfish commented:
Could this be moved along? I don't have a workaround for this right now.

Commit:
Add option to specify table env. in LaTeX output

This commit adds a command line switch (`--latex-table-environment`) for
switching between possible table environments. Currently pandoc is
hardcoded to use `longtable`, which allows for multi-page-tables. This
commit adds support for the tabularx environment (allowing columns that
span over the while remaining table-width) and gives the possibility to
easily add new environments in the future.

pajowu commented:
I updated the documentation according to your Remarks and fixed all merge conflicts. The PR should be mergable again

Regarding my motivation: `longtable` inserts empty pages or overflows the page when float objects are on the same page. This makes `longtable` unusable for us

jgm commented on MANUAL.txt:
```diff
@@ -1148,10 +1148,18 @@ Options affecting specific writers {.options}
+
+:   Use the specified table environment when producing LaTeX output.
+    The default is [`longtable`] which allows table to spread over multiple pages.
+    [`tabularx`] allows for paragraph-like columns whose width automatically expand to fill the width of the environment.
```
Lines should be hard-wrapped to fit within 80 columns.

jgm replied:
I'm not sure I quite understand this description: "paragraph-like columns whose width automatically expand to fill the width of the environment."  Maybe better to say something about what should dictate this choice?  For example: If you have long tables that may need to spans multiple pages, use `longtable`.  If you want to create floating tables (and for better compatibility with other floating elements on the page), use `tabularx`.

jgm commented on MANUAL.txt:
```diff
@@ -1599,6 +1607,10 @@ LaTeX variables are used when [creating a PDF].
     section titles in the header.
 
+`table_environment`
+:   The environemnt given by `--latex-table-environment`
```
"environment"

jgm commented on src/Text/Pandoc/App/CommandLineOptions.hs:
```diff
@@ -850,6 +850,17 @@ options =
                      UTF8.hPutStr stdout (usageMessage prg options)
                      exitSuccess ))
                  "" -- "Show help"
+    , Option "" ["latex-table-class"]
```
Shouldn't this be `latex-table-environment`, given the documentation?

jgm commented on src/Text/Pandoc/Options.hs:
```diff
@@ -168,6 +169,11 @@ data ReferenceLocation = EndOfBlock    -- ^ End of block
 
+-- | Class of tables to use in LaTeX output
+data LatexTableEnvironment = Longtable    -- ^ Use longtable for tables
+                     | Tabularx     -- ^ Use tabularx tables
```
This should align as in the other examples in the code.

jgm commented on src/Text/Pandoc/Options.hs:
```diff
@@ -204,7 +210,8 @@ data WriterOptions = WriterOptions
   , writerSyntaxMap         :: SyntaxMap
   , writerPreferAscii       :: Bool           -- ^ Prefer ASCII representations of characters when possible
-  } deriving (Show, Data, Typeable, Generic)
+  , writerLatexTableEnvironment   :: LatexTableEnvironment
```
Note clearly the API change in the commit comment.

jgm commented on src/Text/Pandoc/Writers/LaTeX.hs:
```diff
@@ -822,13 +835,22 @@ getCaption txt = do
          AlignRight   -> "r"
          AlignCenter  -> "c"
-         AlignDefault -> "l"
+         AlignDefault -> if index == 0 then "l" else "X"
```
Why is "l" used for the first column and then "X"?

jgm commented:
Thanks for the update. I've made some comments to your patch.

Currently there are no tests. There should be tests.

Also, your patch assumes that the only differences between the environments lie in the column specs and environment name.  Have you tested things like captions, footnotes in table cells, and so on, to make sure that this is the case?

jgm commented:
tabu scared me off with this note in the documentation:


> The tabu package patches internal commands of many packages and is liable to stop working if those internal interfaces change. Unfortunately the original author appears to be out of contact and the package had not been updated for some years.

Grandement commented:
I would prefer longtabu as well instead of longtable. The definition of longtabu is close to longtable and I guess it is not a big deal to add it with a optional argument to pandoc. The tabu packge is still working well :). With longtabu automatic linebreaks in a cell are possible due to the column X definition which would be a benifit for pandoc.
